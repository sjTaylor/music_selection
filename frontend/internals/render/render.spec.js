/* global beforeAll afterAll */

import path from 'path';
import fs from 'fs-extra';
import Cli from '../cli/cli';
import Render from './render';

const testFolder = path.join(process.cwd(), 'tmp_testRender');
const mainConfig = require('../config.js')();
const testConfig = {
    rootFolder: testFolder
}
const config = Object.assign(mainConfig, testConfig);
const cli = new Cli(config);
const render = new Render(config);
const rootFolder = path.join(
    config.rootFolder, 
    config.appFolder, 
    config.componentsFolder
)
const containerFolder = path.join(
    config.rootFolder, 
    config.appFolder, 
    config.containersFolder
)
const componentName = 'MainComponent';

describe('Test render of a component', () => {
    beforeAll(() => {
        console.info = () => {};

        // Generate essential files
        const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
        const appIndexJs = path.join(rootFolder, 'index.js');
        const appIndexScss = path.join(scssFolder, 'index.scss');
        const componentScss = path.join(scssFolder, 'components.scss');
        const containersScss = path.join(scssFolder, 'containers.scss');
        const indexJsContent = `
    export {
    };`;
        fs.ensureDirSync(rootFolder);
        fs.ensureDirSync(containerFolder);
        fs.ensureDirSync(scssFolder);
        fs.writeFileSync(appIndexJs, indexJsContent);
        fs.createFileSync(componentScss);
        fs.createFileSync(containersScss);
        fs.createFileSync(appIndexScss);
        const folderPath = path.join(rootFolder, componentName);
        cli.createComponent(folderPath, componentName);
    });

    it('Renders a component', () => {
        const renderedComponent = render.renderComponent(componentName);   
        expect(renderedComponent).toMatchSnapshot();
    })

    it('Renders a list of components', () => {
        const listOfComponents = render.renderListing();
        expect(listOfComponents).toMatchSnapshot();
    })

    it('GetComponents returns a list of components', () => {
        const components = render.getComponents();
        expect(components).toHaveLength(1);
        expect(components).toContain(componentName);
    })

    it('GetContainers returns a list of containers', () => {
        const components = render.getContainers();
        expect(components).toHaveLength(0);
    })
    
    it('Renders a servercomponent', () => {
        const component = render.renderServerComponent(componentName);
        expect(component).toMatchSnapshot();
    })

    afterAll(() => {
        cli.deleteComponentFolder(testFolder);
    });
});