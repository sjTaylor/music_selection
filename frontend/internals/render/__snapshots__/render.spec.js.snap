// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render of a component Renders a component 1`] = `
"<!doctype html>
<html class=\\"no-js\\" lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title>React testing ground</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        <link rel=\\"stylesheet\\" href=\\"/devserver.css\\">
        <link rel=\\"stylesheet\\" href=\\"/static/css/index.css\\">
    </head>
    <body onload=\\"document.documentElement.classList.remove('no-js');\\">
        <div class=\\"devserver\\">
            <form class='devserver__themeform' id='devserver__menu'>
                <a href='/' class='devserver__backlink'>Back to listing</a>
                
                <select
                    class='devserver__themeselect'
                    name='componentName'
                    onChange='devFormChangeComponent(this);'
                    >
                    <option disabled>Choose</option>
                    <option disabled>-Containers</option>
                    
                    <option disabled>-Components</option>
                    <option selected>MainComponent</option>
                </select>
            </form>
        </div>
        
        <div id=\\"root\\"><!-- content --></div>

        
        <script src=\\"/devserver.js\\"></script>
        <script src=\\"/static/js/vendor.js\\"></script>
        <script src=\\"/static/js/index.js\\"></script>
        <!-- render -->
    </body>
</html>
"
`;

exports[`Test render of a component Renders a list of components 1`] = `
"<!doctype html>
<html class=\\"no-js\\" lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title>React testing ground</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        <link rel=\\"stylesheet\\" href=\\"/devserver.css\\">
        <link rel=\\"stylesheet\\" href=\\"/static/css/index.css\\">
    </head>
    <body onload=\\"document.documentElement.classList.remove('no-js');\\">
        <div class=\\"devserver\\">
            <form class='devserver__themeform' id='devserver__menu'>
                <a href='/' class='devserver__backlink'>Back to listing</a>
                
                <select
                    class='devserver__themeselect'
                    name='componentName'
                    onChange='devFormChangeComponent(this);'
                    >
                    <option disabled>Choose</option>
                    <option disabled>-Containers</option>
                    
                    <option disabled>-Components</option>
                    <option >MainComponent</option>
                </select>
            </form>
        </div>
        
        <h2 class='devserver__title'>Containers</h2>
        <br />
        <h2 class='devserver__title'>Components</h2>
        <a href=\\"/MainComponent\\">MainComponent</a><br/>
<div id=\\"root\\"></div>
        
        <script src=\\"/devserver.js\\"></script>
        <script src=\\"/static/js/vendor.js\\"></script>
        <script src=\\"/static/js/index.js\\"></script>
        <!-- render -->
    </body>
</html>
"
`;

exports[`Test render of a component Renders a servercomponent 1`] = `
"<!doctype html>
<html class=\\"no-js\\" lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title>React testing ground</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        <link rel=\\"stylesheet\\" href=\\"/devserver.css\\">
        <link rel=\\"stylesheet\\" href=\\"/static/css/index.css\\">
    </head>
    <body onload=\\"document.documentElement.classList.remove('no-js');\\">
        <div class=\\"devserver\\">
            <form class='devserver__themeform' id='devserver__menu'>
                <a href='/' class='devserver__backlink'>Back to listing</a>
                
                <select
                    class='devserver__themeselect'
                    name='componentName'
                    onChange='devFormChangeComponent(this);'
                    >
                    <option disabled>Choose</option>
                    <option disabled>-Containers</option>
                    
                    <option disabled>-Components</option>
                    <option selected>MainComponent</option>
                </select>
            </form>
        </div>
        
        <div id=\\"root\\"><div class=\\"MainComponent\\" data-reactroot=\\"\\">MainComponent</div></div>

        
        <script src=\\"/devserver.js\\"></script>
        <script src=\\"/static/js/vendor.js\\"></script>
        <script src=\\"/static/js/index.js\\"></script>
        <script>
    ReactDOM.hydrate(React.createElement(Components.MainComponent, {}), document.getElementById(\\"root\\"));
</script>
    </body>
</html>
"
`;
